ddtrace/tracer/abandoned_spans.go

```diff
@@ -0,0 +1,190 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016 Datadog, Inc.
+
+package tracer
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/log"
+)
+
+type AbandonedList struct {
+	head *spansList
+	tail *spansList
+}
+
+type spansList struct {
+	head *spanNode
+	tail *spanNode
+	Next *spansList
+}
+
+type spanNode struct {
+	Element *span
+	Next    *spanNode
+}
+
+func (a *AbandonedList) String() string {
+	var sb strings.Builder
+	for e := a.head; e != nil; e = e.Next {
+		fmt.Fprintf(&sb, "%v", e.String())
+	}
+	return sb.String()
+}
+
+func (s *spansList) String() string {
+	var sb strings.Builder
+	for e := s.head; e != nil; e = e.Next {
+		fmt.Fprintf(&sb, "%v", e.String())
+	}
+	return sb.String()
+}
+
+func (s *spanNode) String() string {
+	sp := s.Element
+	if sp == nil {
+		return "[],"
+	}
+	return fmt.Sprintf("[Span Name: %v, Span ID: %v, Trace ID: %v],", sp.Name, sp.SpanID, sp.TraceID)
+}
+
+func (a *AbandonedList) Extend() {
```

Comment:
I recommend some unit tests for the `AbandonedList` methods, to make sure the code for the linked list functionality works as expected.

Reply:
Also, just to verify, will https://pkg.go.dev/container/list work for you instead? (so you don't have to re-invent the wheel)

Reply:
Good point! I'll add in some unit tests.

In terms of `container/list`: when first working on the task, I did a little benchmarking on different data structures and compared `container/list` to maps, slices, and to my own implementation of a linked list. It seemed my own linked lists worked ever so slightly faster (by a few nanoseconds), so we decided to go with that :)

Reply:
Gotcha! A few nanoseconds _might_ be worth it if this was in the critical path for creating/finishing spans, but since it's a side go routine that only needs to run every minute or two, I'm not sure the additional code and complexity is worth a few nanoseconds. We still have your commits, so we can always go back and use it later if we find a big performance effect to this, but I think going with the standard library code for now will be safer and more maintainable in the future.

Reply:
Perfect, sounds good! I think using the standard library would definitely clear up a lot of code with little performance difference (especially since the logging is now at 10 minutes instead of 1 second like before), and would help remove some of the concerns in other comments below. The goroutine _would_ be running every time a span is created or finished, which is why we were concerned about the performance in the first place. But since this is blocked behind a flag, maybe we can be a little lenient with a couple nanoseconds here and there... I can ping the team for their thoughts!